### PR TITLE
✨ introduce LabelChangedPredicate to trigger event on label change

### PR DESCRIPTION
**Pull request by mars1024**
Signed-off-by: Bruce Ma <brucema19901024@gmail.com>

This is a non-breaking feature :sparkles: .

Just as the title, a `LabelChangedPredicate` is introduced to trigger event on label change, if some extra specification information is carried through object's label, this predicate function will be useful.

### CODE REVIEW

File: pkg/predicate/predicate.go [hunk 1]
```diff
@@ -200,6 +200,36 @@ func (AnnotationChangedPredicate) Update(e event.UpdateEvent) bool {
...
+//
+// Controller.Watch(
+//		&source.Kind{Type: v1.MyCustomKind},
+// 		&handler.EnqueueRequestForObject{},
+//		predicate.Or(predicate.GenerationChangedPredicate{}, predicate.LabelChangedPredicate{}))
+//
+// This will be helpful when object's labels is carrying some extra specification information beyond object's spec,
+// and the controller will be triggered if any valid spec change (not only in spec, but also in labels) happens.
+type LabelChangedPredicate struct {
+	Funcs
```
Comment (alvaroaleman) [inline pkg/predicate/predicate.go:216]: Can you use  `Funcs.UpdateFunc` rather than overwriting it?

Comment (mars1024) [reply pkg/predicate/predicate.go:216]: Sorry, maybe I didn't catch your point, I'm just doing the same way like [ResourceVersionChangedPredicate](https://github.com/kubernetes-sigs/controller-runtime/blob/5564be7542e5d99c77118f2faf3fa9664dc965de/pkg/predicate/predicate.go#L121) and [GenerationChangedPredicate](https://github.com/kubernetes-sigs/controller-runtime/blob/5564be7542e5d99c77118f2faf3fa9664dc965de/pkg/predicate/predicate.go#L155), overriding the `Update` method and inherit the others. 

If we want to assign `Funcs.UpdateFunc`, another constructing function will be introduced for this, and I think maybe it's not easy for using.

Comment (alvaroaleman) [reply pkg/predicate/predicate.go:216]: Ah that's how you came up with it. From a user perspective it is just calling a constructor rather than using a type :)

The reason I think it's slightly better is that it might be confusing to have an embedded `UpdateFunc` described as being called for Update events that does not actually get called for Update events :)

File: pkg/predicate/predicate_test.go [hunk 2]
```diff
@@ -14,6 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+//nolint:dupl
```
Comment (alvaroaleman) [inline pkg/predicate/predicate_test.go:17]: Why is this needed? Can we add an explanation comment for posterity and set it on a more granular level than on the whole package?

File: pkg/predicate/predicate.go (hunk 1 above)
Comment (mars1024) [reply pkg/predicate/predicate.go:216]: :) Got your point, but your suggestion is keeping same as the other predicates or change all the predicates to a new way (use a constructor instead of stating a type), but I think the second way maybe a break change for some "legacy" users.
Or maybe we could just use the **constructing** way for the new `LabelChangedPredicate`?

File: pkg/predicate/predicate_test.go (hunk 2 above)
Comment (mars1024) [reply pkg/predicate/predicate_test.go]: comments is added and level is set to single function from whole package

**Comment by mars1024** [comment]
friendly ping @alvaroaleman , could you help to promote this PR? I think all comments have been addressed.

**Review by alvaroaleman — APPROVED** [review]
/label tide/merge-method-squash
/lgtm
/approve
